tokio/src/io/async_read.rs

```diff
@@ -123,7 +123,9 @@ pub trait AsyncRead {
                 // Convert to `&mut [u8]`
                 let b = &mut *(b as *mut [MaybeUninit<u8>] as *mut [u8]);
 
-                ready!(self.poll_read(cx, b))?
+                let n = ready!(self.poll_read(cx, b))?;
+                assert!(n <= b.len(), "Bad AsyncRead implementation, more bytes were reported as read than the buffer can hold");
```

Comment:
This probably should be debug assert.

Writing bad implemention is unlikely to be intentional

Reply:
We need it to be enforced at all times, since it need to be upheld to prevent UB. Similar to bounds checking slices. Conversely though it might be optimized away for many cases!

Reply:
no, we do not. 
If we do not trust the contract, then this trait itself is UB and should be marked as unsafe

Reply:
It specifically needs to be checked here because it's being passed to `advance_mut` below, which itself is an unsafe fn similar to `Vec::set_len`. Consider if this using a slice instead:
```rust
let n = ready!(self.poll_read(cx, &mut buf))?;
let data = &buf[..n];
```
Accessing the data from the slice (`&buf[..n]`) would be similarly bounds checked, even if the implementation of `poll_read` was incorrect.

Reply:
Actually, the [contract of `BufMut::advance_mut`](https://docs.rs/bytes/0.5.3/bytes/trait.BufMut.html#panics) does seem to prevent this scenario:

> ### Panics
> This function may panic if cnt > self.remaining_mut().
> ### Implementer notes
> It is recommended for implementations of advance_mut to panic if cnt > self.remaining_mut(). If the implementation does not panic, the call must behave as if cnt == self.remaining_mut().
> 
> A call with cnt == 0 should never panic and be a no-op.

@Marwes ?

Reply:
It won't panic and the `Vec` passed here will call `reserve` to expand to the correct length. But it will also follow that by calling `set_len` https://docs.rs/bytes/0.5.3/src/bytes/buf/buf_mut.rs.html#999 which then allows uninitialized bytes to be accessed.

Separately, the `advance_mut` implementation seems dubious. `resize(cnt); set_len(len + cnt);` will always result in the `Vec` exposing undefined values to safe callers. Panicking seems like the better choice there as well.